mock/mock.go

```diff
@@ -598,25 +606,52 @@ func (f argumentMatcher) Matches(argument interface{}) bool {
 	}
 	if argType == nil || argType.AssignableTo(expectType) {
 		result := f.fn.Call([]reflect.Value{arg})
-		return result[0].Bool()
+
+		var matchError error
+		switch {
+		case result[0].Type().Kind() == reflect.Bool:
```

Comment:
Would it make the code simpler if we changed to `if ... else if ... else if ... else` here? Reason being that 'target-less' switch statements aren't obvious in their order of execution if there is more than 1 possible match.